### PR TITLE
test: add tests as well as the ability to run test via `$ pytest` in terminal

### PR DESCRIPTION
**Pull request by aikia-code**
add test for board positions; add test for `check_victory` and `check_defeat`

#4

### CODE REVIEW

File: py_cui_2048/__init__.py
Comment (aikia-code) [inline py_cui_2048/__init__.py:230]: i get a lot of warnings for this. mostly has to do with the use of escape characters `\ `

File: tests/test_2048.py
Comment (jwlodek) [inline tests/test_2048.py]: Instead of doing it in a class, I'd suggest having the test functions just being static, and for constants like `initial_placement` or `game_instance` I'd take a look at pytest fixtures.

If you'd like an example, have a look at the `py_cui` test cases: https://github.com/jwlodek/py_cui/blob/master/tests/conftest.py

And an example of using fixtures in the tests: https://github.com/jwlodek/py_cui/blob/master/tests/test_core/test_grid.py

**Review by jwlodek — CHANGES_REQUESTED** [review]

File: py_cui_2048/__init__.py
Comment (jwlodek) [reply py_cui_2048/__init__.py:230]: Which formatter are you using that's giving you the warnings?

File: tests/test_2048.py
Comment (aikia-code) [reply tests/test_2048.py]: Yes, I chose this approach because it was much more readable. I can definitely include those configurations

File: py_cui_2048/__init__.py
Comment (aikia-code) [reply py_cui_2048/__init__.py:230]:
```cmd
__init__.py:237: SyntaxWarning: invalid escape sequence '\|'
  out = out + "/ __  \|  _  |  /   ||  _  |\n"
__init__.py:238: SyntaxWarning: invalid escape sequence '\ '
  out = out + "`' / /'| |/' | / /| | \ V / \n"
__init__.py:239: SyntaxWarning: invalid escape sequence '\ '
  out = out + "  / /  |  /| |/ /_| | / _ \ \n"
__init__.py:240: SyntaxWarning: invalid escape sequence '\ '
  out = out + "./ /___\ |_/ /\___  || |_| |\n"
__init__.py:241: SyntaxWarning: invalid escape sequence '\_'
  out = out + "\_____/ \___/     |_/\_____/\n"
```

My environment is python `v3.12.1`

I have not used any of my linters and formatters on your code yet.

i mainly use **Black and Pylint**

**Comment by aikia-code** [comment]
There is indeed some performance gains in using test functions over classes

**Comment by ProsperoKay** [comment]
There is indeed some perfomance gains in using test functions over classes

File: tests/test_2048.py
Comment (jwlodek) [inline tests/test_2048.py]: This is actually catching unintended behavior - I would think `check_victory` should return `False` if victory is not found. Could you change the underlying method to do this and update this test case?

**Review by jwlodek — CHANGES_REQUESTED** [review]

**Comment by jwlodek** [comment]
Other than that this looks good to me, I will merge once `check_victory` is modified.

File: py_cui_2048/__init__.py
Comment (jwlodek) [reply py_cui_2048/__init__.py:230]: We can probably get around these warnings by using a raw string.

Comment (aikia-code) [reply py_cui_2048/__init__.py:230]: Yes. I realised py_cui reads the ascii characters directly. i thought escaping those characters might work, but i think it may result in unintended results. I can tackle that once i am able to run the entire program in my environment. Using raw string should work

File: tests/test_2048.py
Comment (aikia-code) [reply tests/test_2048.py]: i thought the same.

Comment (aikia-code) [reply tests/test_2048.py]: After merge i would have to get the program running in my environment to do some integration tests

Comment (aikia-code) [reply tests/test_2048.py]: I Just don't want to make any breaking changes to your code just yet

**Review by jwlodek — APPROVED** [review]

**Pull request merged**
